### PR TITLE
Improve type annotations for `jit.script`

### LINKED ISSUE #108781: Improve type annotations for `jit.script`
### 🚀 The feature, motivation and pitch

Looking at the code https://github.com/pytorch/pytorch/blob/c75aec90d3e901abd604d075c5695ae561608605/torch/jit/_script.py#L1068-L1391,

I derived the type-hint unit test

```python
from enum import Enum
from typing import Type

from torch import jit, nn, ScriptDict, ScriptFunction, ScriptList

from typing_extensions import assert_never, assert_type


class Color(Enum):
    RED = 1
    GREEN = 2
    BLUE = 3


# Script Enum
assert_type(jit.script(Color), Type[Color])

# ScriptDict
assert_type(jit.script({1: 1}), ScriptDict)

# ScriptList
assert_type(jit.script([0]), ScriptList)

# ScriptModule
assert_type(jit.script(nn.Linear(2, 2)), jit.RecursiveScriptModule)

# ScripFunction
assert_type(jit.script(nn.functional.relu), ScriptFunction)

# can't script nn.Module class
assert_never(jit.script(nn.Linear))
```

However, this currently fails, here's the actual output from `mypy`:

```
bugs/jit_script_type_hints.py:4:1:4:1: error: Module "torch" has no attribute "ScriptDict"  [attr-defined]
    from torch import ScriptDict, ScriptFunction, ScriptList, jit, nn
    ^
bugs/jit_script_type_hints.py:4:1:4:1: error: Module "torch" has no attribute "ScriptList"  [attr-defined]
    from torch import ScriptDict, ScriptFunction, ScriptList, jit, nn
    ^
bugs/jit_script_type_hints.py:15:1:15:1: error: Expression is of type "Any", not "type[Color]"  [assert-type]
    assert_type(jit.script(Color), Type[Color])
    ^
bugs/jit_script_type_hints.py:24:1:24:1: error: Expression is of type "Any", not "RecursiveScriptModule"  [assert-type]
    assert_type(jit.script(nn.Linear(2, 2)), jit.RecursiveScriptModule)
    ^
bugs/jit_script_type_hints.py:27:1:27:1: error: Expression is of type "Any", not "ScriptFunction"  [assert-type]
    assert_type(jit.script(nn.functional.relu), ScriptFunction)
    ^
Found 5 errors in 1 file (checked 1 source file)
```

### Alternatives

_No response_

### Additional context

_No response_

cc @EikanWang @jgong5 @wenzhe-nrv @sanchitintel

### PR DESCRIPTION
**Pull request by randolf-scholz**
Fixes #108781

- [x] added `@overload` for `jit.script`
- [x] added typing unittest in `test/typing/pass/jit.py`
    - NOTE: unittest is not automatically checked by mypy when executing lintrunner currently. (how to fix?)
- [x] used `stubgen` to create [torch/jit/_script.pyi](https://github.com/pytorch/pytorch/pull/108782/files#diff-738e66abee2523a952b3ddbaecf95e187cce559473cf8c1b3da7c247ee5d1132) and added overloads there. (adding them inside `_script.py` itself interfered with JIT engine)

cc @ezyang @malfet @rgommers @xuzhao9 @gramster

### CODE REVIEW

**Review by ezyang — APPROVED** [review]

**Comment by randolf-scholz** [comment]
@ezyang The local lintrunner did not catch some typing errors caught by the CI, it seems they run MYPY with different configs.
I patched the code, by adding `ParamSpec` to `ScriptFunction` and `jit.script`

File: torch/distributed/optim/optimizer.py [hunk 1]
```diff
@@ -236,10 +236,11 @@ def step(self, context_id):
...
 
-        if self.is_functional_optim:
-            optimizer_step_func = _script_local_optimizer_step
-        else:
-            optimizer_step_func = _local_optimizer_step
+        optimizer_step_func = (
+            _script_local_optimizer_step
+            if self.is_functional_optim
+            else _local_optimizer_step
+        )
```
Comment (randolf-scholz) [inline torch/distributed/optim/optimizer.py:243]: This is a trick to get mypy to use `Union` of either branch

File: torch/jit/_script.py [hunk 2]
```diff
@@ -1065,11 +1079,95 @@ def create_script_list(obj, type_hint=None):
...
+
+@overload
+def script(  # type: ignore[misc]
+    obj: Callable[P, R],
+    optimize: Optional[bool] = None,
+    _frames_up: int = 0,
+    _rcb: Optional[ResolutionCallback] = None,
+    example_inputs: Union[List[Tuple], Dict[Callable, List[Tuple]], None] = None,
+) -> ScriptFunction[P, R]:
+    ...
```
Comment (randolf-scholz) [inline torch/jit/_script.py:1152]: Added ParamSpec which ensures `mypy` recognizes that jit.script doesn't modify the signature of the callable.

File: torch/jit/_script.py [hunk 3]
```diff
@@ -1065,11 +1079,95 @@ def create_script_list(obj, type_hint=None):
     return torch._C.ScriptList(obj)  # type: ignore[attr-defined]
 
 
+# Defined in torch/csrc/jit/python/script_init.cpp
+ResolutionCallback = Callable[[str], Callable[..., Any]]
+ClassVar = TypeVar("ClassVar", bound=type)
+P = ParamSpec("P")
+R = TypeVar("R", covariant=True)  # return value
```
Comment (randolf-scholz) [inline torch/jit/_script.py:1086]: Should these be moved to the top, after imports?

**Comment by ezyang** [comment]
seems you still have to work it more

**Comment by ezyang** [comment]
seems to pass ci lol. you still working it?

**Comment by randolf-scholz** [comment]
@ezyang The earlier version passed CI, but CI wasn't testing some things. 

For instance, one couldn't specialize to `ScriptList[int]`. I found a way to fix this by adding `__class_getitem__` to the C++ class. However, this will never work for python 3.8 since it requires https://peps.python.org/pep-0585/, which only became available with 3.9.

So, currently I am thinking of reverting some of these changes, so that it remains compatible with 3.8.

**Comment by ezyang** [comment]
ok just ping me when you're ready

**Comment by randolf-scholz** [comment]
@ezyang I think this is good now, I basically removed Generic functionality, this can be added in the future when `pytorch` deprecates python3.8.

The two CI failures seem like server issues, can you restart these jobs? (or can I do it somehow?)

**Review by ezyang — APPROVED** [review]
Yeah, I'm not sure how to setup mypy-primer :/

**Comment by randolf-scholz** [comment]
@ezyang Not sure what happened, but it seems it used the wrong commit? I merged upstream and reapplied the changes.

**Comment by randolf-scholz** [comment]
@ezyang Hey, it seems like some of my type hint contributions are shipping with 2.2, but this one is missing, it seems like the new stub file `_script.pyi` is not included in the wheels.

Does it need to be marked as included somewhere?

**Comment by ezyang** [comment]
The directory containing the pyi file has to be manually listed in setup.py so that is probably the problem. Send us a PR

**Comment by randolf-scholz** [comment]
@ezyang Completely forgot about this, here is the PR: #129654

**Pull request closed without merging**